lib/std/x/os/socket_posix.zig

```diff
@@ -0,0 +1,235 @@
+    pub fn setLinger(self: Socket, timeout_seconds: ?u16) !void {
+        if (comptime @hasDecl(os, "SO_LINGER")) {
+            const settings = extern struct {
+                l_onoff: c_int,
+                l_linger: c_int,
+            }{
+                .l_onoff = @intCast(c_int, @boolToInt(timeout_seconds != null)),
+                .l_linger = if (timeout_seconds) |seconds| @intCast(c_int, seconds) else 0,
+            };
+
+            return self.setOption(os.SOL_SOCKET, os.SO_LINGER, mem.asBytes(&settings));
+        }
+
+        return error.UnsupportedSocketOption;
+    }
+
+    /// On connection-oriented sockets, have keep-alive messages be sent periodically. The timing in which keep-alive
+    /// messages are sent are dependant on operating system settings. It returns `error.UnsupportedSocketOption` if
+    /// the host does not support periodically sending keep-alive messages on connection-oriented sockets. 
+    pub fn setKeepAlive(self: Socket, enabled: bool) !void {
+        if (comptime @hasDecl(os, "SO_KEEPALIVE")) {
+            return self.setOption(os.SOL_SOCKET, os.SO_KEEPALIVE, mem.asBytes(&@as(u32, @boolToInt(enabled))));
+        }
+        return error.UnsupportedSocketOption;
+    }
+
+    /// Allow multiple sockets on the same host to listen on the same address. It returns `error.UnsupportedSocketOption` if
+    /// the host does not support sockets listening the same address.
+    pub fn setReuseAddress(self: Socket, enabled: bool) !void {
+        if (comptime @hasDecl(os, "SO_REUSEADDR")) {
+            return self.setOption(os.SOL_SOCKET, os.SO_REUSEADDR, mem.asBytes(&@as(u32, @boolToInt(enabled))));
+        }
+        return error.UnsupportedSocketOption;
+    }
+
+    /// Allow multiple sockets on the same host to listen on the same port. It returns `error.UnsupportedSocketOption` if
+    /// the host does not supports sockets listening on the same port.
+    pub fn setReusePort(self: Socket, enabled: bool) !void {
+        if (comptime @hasDecl(os, "SO_REUSEPORT")) {
+            return self.setOption(os.SOL_SOCKET, os.SO_REUSEPORT, mem.asBytes(&@as(u32, @boolToInt(enabled))));
+        }
+        return error.UnsupportedSocketOption;
+    }
+
+    /// Set the write buffer size of the socket.
+    pub fn setWriteBufferSize(self: Socket, size: u32) !void {
+        return self.setOption(os.SOL_SOCKET, os.SO_SNDBUF, mem.asBytes(&size));
+    }
+
+    /// Set the read buffer size of the socket.
+    pub fn setReadBufferSize(self: Socket, size: u32) !void {
+        return self.setOption(os.SOL_SOCKET, os.SO_RCVBUF, mem.asBytes(&size));
+    }
+
+    /// Set a timeout on the socket that is to occur if no messages are successfully written
+    /// to its bound destination after a specified number of milliseconds. A subsequent write
+    /// to the socket will thereafter return `error.WouldBlock` should the timeout be exceeded.
+    pub fn setWriteTimeout(self: Socket, milliseconds: usize) !void {
```

Comment:
I would put a big warning on these functions: they are incompatible with non-blocking sockets.
I'd almost prefer we didn't even provide them as they're a footgun for most users.

Reply:
Warning added.